inference-engine/src/mkldnn_plugin/nodes/mkldnn_detection_output_node.cpp

```diff
@@ -125,480 +152,753 @@ void MKLDNNDetectionOutputNode::initSupportedPrimitiveDescriptors() {
+                        maxCIdx = c;
+                    }
+                }
+                if (maxCIdx > 0) {
+                    // include this prior and class with max conf
+                    mtx.lock();
+                    indicesData[offB + detectionsData[n*classesNum]] = maxCIdx*priorsNum + p;  // de-refer to get prior and class id.
+                    detectionsData[n*classesNum]++;
+                    mtx.unlock();
+                }
+            }
+        });
+
+        // in:  pindices, detectionCount(filtered num)
+        // out: buffer, detectionCount(k)
+        int count = detectionsData[n*classesNum];
+        int k = (topK == -1 ? count : (std::min)(topK, count));
+
+        const float *pconf = reorderedConfData + offB;
+        int *indices = indicesData + offB;
+        int *pbuffer = indicesBufData + offB;
+        topk(indices, pbuffer, pconf, count, k);
+        detectionsData[n * classesNum] = k;
     }
+}
 
-    float intersect_xmin = (std::max)(xmin1, xmin2);
-    float intersect_ymin = (std::max)(ymin1, ymin2);
-    float intersect_xmax = (std::min)(xmax1, xmax2);
-    float intersect_ymax = (std::min)(ymax1, ymax2);
-
-    float intersect_width  = intersect_xmax - intersect_xmin;
-    float intersect_height = intersect_ymax - intersect_ymin;
-
-    if (intersect_width <= 0 || intersect_height <= 0) {
-        return 0.0f;
+inline void MKLDNNDetectionOutputNode::confReorderAndFilterSparsityCF(const float* confData, const float* ARMConfData, float* reorderedConfData,
+    int* indicesData, int* indicesBufData, int* detectionsData) {
+    int* reorderedConfDataIndices = reinterpret_cast<int*>(reorderedConfData);
+    for (int n = 0; n < imgNum; ++n) {
+        int off = n * priorsNum * classesNum;
+        int offV = n * priorsNum;  // vertical info
+
+        int offH = n * confInfoLen * classesNum; // horizontal info
+        // reset count
+        parallel_for(classesNum, [&](size_t c) {
+            int countIdx = offH + c * confInfoLen + priorsNum;
+            reorderedConfDataIndices[countIdx] = 0;
+        });
+
+        std::mutex mtx;
+        parallel_for(numPriorsActual[n], [&](size_t p) {
+            // intentionally code branch from higher level
+            if (withAddBoxPred) {
+                bool isARMPrior = ARMConfData[n * priorsNum * 2 + p * 2 + 1] < objScore;
+                bool priorStatusSet = false;
+                if (isShareLoc)
+                    confInfoForPrior[offV + p] = -1;
+                int confIdxPrior = off + p * classesNum;
+                for (int c = 0; c < classesNum; ++c) {
+                    float conf = confData[confIdxPrior + c];
+                    if (isARMPrior)
+                        conf = (c == backgroundClassId) ? 1.0f : 0.0f;
+                    if (conf > confidenceThreshold) {
+                        int idx = offH + c * confInfoLen;
+                        reorderedConfData[idx + p] = conf;
+                        mtx.lock();
```

Comment:
Have you tried to avoid mutex here? Like changing order of loops... Is this part of code bottleneck at all?

Reply:
Each prior in parallel process have the possibility to change the statistic info for all classes. It's hard to segment to safe access. Meanwhile, here mutex(sparsity lock check) does not introduce noticeable perf penalty given here is still memory intensive case, although already mitigated dramatically than dense case. Thank you!